packages/event-handler/src/rest/middleware/compress.ts

```diff
@@ -0,0 +1,120 @@
+import type { Middleware } from '../../types/index.js';
+import type { CompressionOptions } from '../../types/rest.js';
+import {
+  CACHE_CONTROL_NO_TRANSFORM_REGEX,
+  COMPRESSIBLE_CONTENT_TYPE_REGEX,
+  COMPRESSION_ENCODING_TYPES,
+} from '../constants.js';
+
+/**
+ * Compresses HTTP response bodies using standard compression algorithms.
+ *
+ * This middleware automatically compresses response bodies when they exceed
+ * a specified threshold and the client supports compression. It respects
+ * cache-control directives and only compresses appropriate content types.
+ *
+ * The middleware checks several conditions before compressing:
+ * - Response is not already encoded or chunked
+ * - Request method is not HEAD
+ * - Content length exceeds the threshold
+ * - Content type is compressible
+ * - Cache-Control header doesn't contain no-transform
+ * - Response has a body
+ *
+ * **Basic compression with default settings**
+ *
+ * @example
+ * ```typescript
+ * import { Router } from '@aws-lambda-powertools/event-handler';
+ * import { compress } from '@aws-lambda-powertools/event-handler/rest/middleware';
+ *
+ * const app = new Router();
+ *
+ * app.use(compress());
+ *
+ * app.get('/api/data', async () => {
+ *   return { data: 'large response body...' };
+ * });
+ * ```
+ *
+ * **Custom compression settings**
+ *
+ * @example
+ * ```typescript
+ * import { Router } from '@aws-lambda-powertools/event-handler';
+ * import { compress } from '@aws-lambda-powertools/event-handler/rest/middleware';
+ *
+ * const app = new Router();
+ *
+ * app.use(compress({
+ *   threshold: 2048,
+ *   encoding: 'deflate'
+ * }));
+ *
+ * app.get('/api/large-data', async () => {
+ *   return { data: 'very large response...' };
+ * });
+ * ```
+ *
+ * @param options - Configuration options for compression behavior
+ * @param options.threshold - Minimum response size in bytes to trigger compression (default: 1024)
+ * @param options.encoding - Preferred compression encoding to use when client supports multiple formats
+ */
+
+const compress = (options?: CompressionOptions): Middleware => {
+  const threshold = options?.threshold ?? 1024;
+
+  return async (_, reqCtx, next) => {
+    await next();
```

Comment:
The [Python docs](https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#compress) suggest that even with `compress` enabled, if the client doesn't send the correct `Accept-Encoding` header, Event Handler won't return a compressed response.

Can you check their implementation to see if this is this the case?

Reply:
@svozza what do you think about this? I saw at least another JS implementation doesn't do this, but I think the client should be able to bypass the API compression setting if they don't want to/can't support it, otherwise the request would be thrown away anyway.

This also matches the behavior from API Gateway as documented [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-make-request-with-compressed-payload.html) and [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-gzip-compression-decompression.html).

Reply:
No I agree, that header is there for a reason and we should respect it.